### PR TITLE
[inference/model]Adapted to the baichuan2-7B model

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
pytest:
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/config.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 
 _DEFAULT_PROMPT_TEMPLATES = {
     "llama": "[INST] <<SYS>>\nYou are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe.  Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature. If a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\n<</SYS>>\n{input_text}[/INST]",
+    "baichuan": "<reserved_106>{input_text}<reserved_107>",
```
Comment (yuanheng-zhao) [inline colossalai/inference/config.py:29]: Where the prompt template is from?

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+from colossalai.inference.flash_decoding_utils import FDIntermTensors
+from colossalai.inference.modeling.models.nopadding_llama import llama_base_attn_forward
+from colossalai.kernel.kernel_loader import InferenceOpsLoader
+from colossalai.logging import get_dist_logger
+
+inference_ops = InferenceOpsLoader().load()
+
+logger = get_dist_logger(__name__)
+
+# def get_alibi_slopes(n_head):
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_baichuan.py:16]: Unimplemented or forgot-to-remove

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 3]
```diff
@@ -370,113 +369,23 @@ def forward(
...
+            cos_sin=cos_sin,
+            fd_inter_tensor=fd_inter_tensor,
+            is_prompts=is_prompts,
+            kv_seq_len=kv_seq_len,
+            output_tensor=output_tensor,
+            sm_scale=sm_scale,
+            use_cuda_kernel=use_cuda_kernel,
+            cu_seqlens=cu_seqlens,
+            high_precision=high_precision,
+        )
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_llama.py:388]: Unnecessary operation

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 4]
```diff
@@ -534,3 +443,152 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
+            output=output_tensor,
+            mid_output=fd_inter_tensor.mid_output,
+            mid_output_lse=fd_inter_tensor.mid_output_lse,
+            sm_scale=sm_scale,
+            q_len=q_len,
+        )
+
+    attn_output = attn_output.view(-1, self.hidden_size)
+    attn_output = torch.mm(attn_output, self.o_proj_weight)
+
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_llama.py:593]: Do not extract forward function

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+
+class NopadBaiChuanAttention(nn.Module):
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_baichuan.py:19]: Exactly the same as `NopadLlamaAttention` in `nopadding_llama.py`. Why not just replace as llama layer in baichuan policy, then we don't need to put redundant layers in modeling

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            output_tensor=output_tensor,
+            sm_scale=sm_scale,
+            use_cuda_kernel=use_cuda_kernel,
+            cu_seqlens=cu_seqlens,
+            high_precision=high_precision,
+        )
+
+
+# NOTE This will cause the result to be different from the transformer in some cases.
+class NopadBaichuanMLP(nn.Module):
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_baichuan.py:135]: Same issue as above. Exactly the same as `NopadLlamaMLP`

File: colossalai/inference/modeling/policy/nopadding_baichuan.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    llama_rmsnorm_forward,
+)
+from colossalai.inference.utils import init_to_get_rotary
+from colossalai.shardformer.policies.base_policy import ModulePolicyDescription, SubModuleReplacementDescription
+
+# import colossalai
+from colossalai.shardformer.policies.llama import LlamaForCausalLMPolicy
+
+
+class NoPaddingBaiChuanModelInferPolicy(LlamaForCausalLMPolicy):
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/nopadding_baichuan.py:18]: Consider the comment left in baichuan modeling

File: tests/test_infer/test_models/model_utils/baichuan2_7B/modeling_baichuan.py [hunk 8]
```diff
@@ -0,0 +1,801 @@
+# Copyright 2023 Baichuan Inc. All Rights Reserved.
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_models/model_utils/baichuan2_7B/modeling_baichuan.py:1]: Consider other ways of testing. I know you would like to compare the outputs of original model and the adapted one, however, moving the whole (original) model here only for the reason of testing is pretty weird and unnecessary.

File: tests/test_infer/test_models/model_utils/baichuan2_7B/generation_utils.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
+from queue import Queue
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_models/model_utils/baichuan2_7B/generation_utils.py:1]: Same as above. Remove original model and config

File: tests/test_infer/test_models/model_utils/baichuan2_7B/configuration_baichuan.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Baichuan Inc. All Rights Reserved.
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_models/model_utils/baichuan2_7B/configuration_baichuan.py:1]: Same as above. Remove config

File: tests/test_infer/test_models/test_baichuan.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        )
+    ).cuda()
+    model = model.eval()
+
+    inputs = [
+        "介绍一下今天的北京,比如故宫，天安门，长城或者其他的一些景点,",
+    ]
+
+    output_len = 38
+    do_sample = True
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_models/test_baichuan.py:39]: Why `do_sample` set to `True`? You're not testing on greedy search

File: colossalai/inference/config.py (hunk 1 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/config.py:29]: This is based on the results printed by baichuan2 7b chat.

File: colossalai/inference/modeling/models/nopadding_baichuan.py (hunk 2 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:16]: OK, I will remove it.

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 3 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 4 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

File: colossalai/inference/modeling/models/nopadding_baichuan.py (hunk 5 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:17]: Done

File: colossalai/inference/modeling/policy/nopadding_baichuan.py (hunk 7 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/policy/nopadding_baichuan.py:18]: Done

File: colossalai/inference/modeling/models/nopadding_baichuan.py (hunk 6 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/modeling/models/nopadding_baichuan.py:137]: Done

File: tests/test_infer/test_models/test_baichuan.py (hunk 11 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_models/test_baichuan.py]: Done

File: tests/test_infer/test_models/model_utils/baichuan2_7B/configuration_baichuan.py (hunk 10 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_models/model_utils/baichuan2_7B/configuration_baichuan.py]: Done

File: tests/test_infer/test_models/model_utils/baichuan2_7B/generation_utils.py (hunk 9 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_models/model_utils/baichuan2_7B/generation_utils.py]: Done

File: tests/test_infer/test_models/model_utils/baichuan2_7B/modeling_baichuan.py (hunk 8 above)
Comment (yuehuayingxueluo) [reply tests/test_infer/test_models/model_utils/baichuan2_7B/modeling_baichuan.py]: Done

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+from colossalai.inference.modeling.models.nopadding_llama import NopadLlamaAttention
+from colossalai.kernel.kernel_loader import InferenceOpsLoader
+from colossalai.logging import get_dist_logger
+
+inference_ops = InferenceOpsLoader().load()
+
+logger = get_dist_logger(__name__)
+
+
+class NopadBaiChuanAttention(nn.Module):
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_baichuan.py:17]: trivial: Baichuan, caps to be consistent with `NopadBaichuanMLP`

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        self.head_dim = self.hidden_size // self.num_heads
+
+        # Used to adapt llama_base_attn_forward
+        self.num_key_value_heads = self.num_heads
+
+        qkv_weight_list = [attn_qproj_w, attn_kproj_w, attn_vproj_w]
+        self.qkv_weight = torch.stack(qkv_weight_list, dim=0)
+
+    @staticmethod
+    def from_native_module(module: nn.Module, *args, **kwargs) -> nn.Module:
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_baichuan.py:50]: replace `nn.Module`

File: colossalai/inference/modeling/models/nopadding_baichuan.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        """Used for initialize the weight of NopadBaiChuanAttention by origin BaiChuanAttention.
+
+        Args:
+            module (BaiChuanAttention): The origin BaiChuanAttention layer.
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/models/nopadding_baichuan.py:54]: trivial: The original layer is `BaichuanAttention`. You could just use `module (nn.Module)` here in docstring

File: colossalai/inference/modeling/policy/nopadding_baichuan.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+from colossalai.inference.modeling.models.nopadding_llama import (
+    llama_causal_lm_forward,
+    llama_decoder_layer_forward,
+    llama_model_forward,
+    llama_rmsnorm_forward,
+)
+from colossalai.inference.utils import init_to_get_rotary
+from colossalai.shardformer.policies.base_policy import ModulePolicyDescription, SubModuleReplacementDescription
+
+# import colossalai
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/nopadding_baichuan.py:14]: remove unused import

File: colossalai/inference/modeling/policy/__init__.py [hunk 16]
```diff
@@ -1,9 +1,16 @@
 from .glide_llama import GlideLlamaModelPolicy
+from .nopadding_baichuan import NoPaddingBaiChuanModelInferPolicy
 from .nopadding_llama import NoPaddingLlamaModelInferPolicy
 
 model_policy_map = {
     "nopadding_llama": NoPaddingLlamaModelInferPolicy,
+    "nopadding_baichuan": NoPaddingBaiChuanModelInferPolicy,
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/__init__.py:7]: Trivial: Consistency in caps `Baichuan`

File: tests/test_infer/test_models/test_baichuan.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+from colossalai.inference.flash_decoding_utils import FDIntermTensors
+from colossalai.testing import parameterize, rerun_if_address_is_in_use, spawn
+
+PATH_EXIST = "baichuan-inc/Baichuan2-7B-Base"
+
+if os.path.exists(PATH_EXIST):
+    PATH_EXIST = True
+else:
+    PATH_EXIST = False
+
```
Comment (yuanheng-zhao) [inline tests/test_infer/test_models/test_baichuan.py:21]: A more explicit way
```python
BAICHUAN_MODEL_NAME_OR_PATH = "baichuan-inc/Baichuan2-7B-Base"
# remove L17~20
...
def check_inference_engine(use_engine=False, prompt_template=None):
    ...
    tokenizer = AutoTokenizer.from_pretrained(BAICHUAN_MODEL_NAME_OR_PATH, use_fast=False, trust_remote_code=True)
    model = AutoModelForCausalLM.from_pretrained(
        BAICHUAN_MODEL_NAME_OR_PATH, device_map="auto", torch_dtype=torch.bfloat16, trust_remote_code=True
    ).cuda()

...

@pytest.mark.skipif(
    os.path.exists(BAICHUAN_MODEL_NAME_OR_PATH),
    reason="There is no local model address included, please replace this address with a valid one.",
)
...


```

**Review by yuanheng-zhao — APPROVED** [review]

**Pull request merged**